src/DataSet/ObjectDataSet.php

```diff
@@ -218,17 +225,41 @@ public function getRules(): iterable
         if ($this->rulesProvided) {
             /** @var RulesProviderInterface $object */
             $object = $this->object;
-
-            return $object->getRules();
+            $rules = $object->getRules();
+        } else {
+            $rules = [];
         }
 
-        // Providing data set assumes object has its own rules getting logic. So further parsing of rules is skipped
-        // intentionally.
+        // Providing data set assumes object has its own rules getting logic.
+        // So further parsing of rules is skipped intentionally.
         if ($this->dataSetProvided) {
-            return [];
+            return $rules;
+        }
+
+        // Merge rules from `RulesProviderInterface` implementation and parsed from PHP attributes.
+        $rules = $rules instanceof Traversable ? iterator_to_array($rules) : $rules;
+        foreach ($this->parser->getRules() as $key => $value) {
+            if (is_int($key)) {
+                array_unshift($rules, $value);
+                continue;
+            }
+
+            /**
+             * @psalm-var list<RuleInterface> $value If `$key` is string, then `$value` is array of rules
+             * @see ObjectParser::getRules()
+             */
+
+            if (!isset($rules[$key])) {
+                $rules[$key] = $value;
+                continue;
+            }
+
+            $rules[$key] = is_iterable($rules[$key])
```

Comment:
If `$rules[$key]` isn't iterable, why does it mean that `$value` is?

Btw, this line must be benchmarked. I'd suggest to use `foreach` loop with `[]=` operator

Reply:
> If `$rules[$key]` isn't iterable, why does it mean that `$value` is?

`$value` is always list of rules here. See https://github.com/yiisoft/validator/blob/master/src/Helper/ObjectParser.php#L197

> Btw, this line must be benchmarked. I'd suggest to use foreach loop with []= operator

https://github.com/vjik/php-unpacking-vs-foreach-bench

I think in this case better use unpacking.